### PR TITLE
main: FieldValue interface

### PR DESCRIPTION
**Pull request by lavrukov**

### CODE REVIEW

**Review by nvamelichev — CHANGES_REQUESTED** [review]

File: databind/src/main/java/tech/ydb/yoj/databind/expression/values/BoolFieldValue.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package tech.ydb.yoj.databind.expression.values;
+
+import tech.ydb.yoj.databind.FieldValueType;
+import tech.ydb.yoj.databind.expression.FieldValue;
+
+import java.lang.reflect.Type;
+
+public record BoolFieldValue(
```
[the remark this answers is not in the document]
Comment (lavrukov) [reply databind/src/main/java/tech/ydb/yoj/databind/expression/values/BoolFieldValue.java:8]: It's good Idea, but I don't know how to solve name collision like ByteArray, and Tuple

**Comment by nvamelichev** [comment]
@lavrukov Please fix minor checkstyle failures:
```
Error:  /home/runner/work/yoj-project/yoj-project/databind/src/main/java/tech/ydb/yoj/databind/expression/values/RealFieldValue.java:13:36: 'switch' is not preceded with whitespace. [WhitespaceAround]
Error:  /home/runner/work/yoj-project/yoj-project/databind/src/main/java/tech/ydb/yoj/databind/expression/values/TimestampFieldValue.java:14:36: 'switch' is not preceded with whitespace. [WhitespaceAround]
Error:  /home/runner/work/yoj-project/yoj-project/databind/src/main/java/tech/ydb/yoj/databind/expression/values/NumberFieldValue.java:14:36: 'switch' is not preceded with whitespace. [WhitespaceAround]
Error:  /home/runner/work/yoj-project/yoj-project/databind/src/main/java/tech/ydb/yoj/databind/expression/values/FieldValue.java:37:1: '{' at column 1 should be on the previous line. [LeftCurly]
Error:  /home/runner/work/yoj-project/yoj-project/databind/src/main/java/tech/ydb/yoj/databind/expression/values/StringFieldValue.java:16:36: 'switch' is not preceded with whitespace. [WhitespaceAround]
```

and this will be ready to merge, I think :-)

**Comment by nvamelichev** [comment]
TODO: Move FieldValue interface a package up (to where it was) for better compatibility

**Comment by nvamelichev** [comment]
**UPD:** Brought back compatibility `FieldValue` interface in old package (with deprecation warnings, of course!)

**Comment by nvamelichev** [comment]
@lavrukov Sasha, please look at this. I've made the `FieldValue` interface a bit more backwards-compatible then before, for a smooth transition :-)

(And we'll still release this as the next major version, 2.**6**.0)

File: databind/src/main/java/tech/ydb/yoj/databind/expression/values/BooleanFieldValue.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package tech.ydb.yoj.databind.expression.values;
```
Comment (lavrukov) [inline databind/src/main/java/tech/ydb/yoj/databind/expression/values/BooleanFieldValue.java:1]: nit: maybe name package like `fieldvalue` ?

File: databind/src/main/java/tech/ydb/yoj/databind/expression/ModelField.java [hunk 3]
```diff
@@ -67,25 +76,25 @@ public Stream<ModelField> flatten() {
 
     @NonNull
     public FieldValue validateValue(@NonNull FieldValue value) {
-        if (value.isTuple()) {
-            value.getTuple().streamComponents()
+        if (value instanceof TupleFieldValue tupleValue) {
```
Comment (lavrukov) [inline databind/src/main/java/tech/ydb/yoj/databind/expression/ModelField.java:79]: logic under instanceof for interface looks little bit strange. Maybe it possible to make part of interface?
```
if(!value.check(fieldValueType)) {
    if (userFieldPath != null) {
         throw value.buildUserException(userFieldPath)
    } else {
        throw new IllegalArgumentException("formatted message")
    }
}

```

File: databind/src/main/java/tech/ydb/yoj/databind/expression/values/ByteArrayFieldValue.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+                    .formatted(valueType, this)
+            );
+        }
+
+        return byteArray;
+    }
+
+    @Override
+    public String toString() {
+        return "byte array with hash " + hashCode();
```
[the remark this answers is not in the document]
Comment (lavrukov) [reply databind/src/main/java/tech/ydb/yoj/databind/expression/values/ByteArrayFieldValue.java]: we did it beacause there could be a private data, ifiar

File: databind/src/main/java/tech/ydb/yoj/databind/expression/values/NumberFieldValue.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
+package tech.ydb.yoj.databind.expression.values;
+
+import tech.ydb.yoj.databind.FieldValueType;
+
+import java.lang.reflect.Type;
+import java.time.Instant;
+import java.util.Optional;
+
+public record NumberFieldValue(
```
[the remark this answers is not in the document]
Comment (nvamelichev) [reply databind/src/main/java/tech/ydb/yoj/databind/expression/values/NumberFieldValue.java]: :heavy_check_mark: renamed, actually

File: databind/src/main/java/tech/ydb/yoj/databind/expression/values/ByteArrayFieldValue.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (nvamelichev) [reply databind/src/main/java/tech/ydb/yoj/databind/expression/values/ByteArrayFieldValue.java]: OK, that's not a big deal, and if someone needs more informative `ByteArrayFieldValue.toString()` we can always reconsider

**Review by nvamelichev — APPROVED** [review]

**Pull request merged**